keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1080 @@
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The current `readOnly` volumes are not recursively read-only, and may result in compromisation of data;
+e.g., even if `/mnt` is mounted as read-only, its submounts such as `/mnt/usbstorage` are not read-only.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+Support recursive read-only mounts for runc >= 1.1 && kernel >= 5.12.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+Support recursive read-only mounts for old runc and old kernel releases.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants to mount `/mnt`, includings its submounts such as `/mnt/usbstorage`, as read-only.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+Constraints: needs runc >= 1.1 && kernel >= 5.12.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risk: a manifest with `recursiveReadOnly: Enabled` will not work on old distros (See "Constraints" above).
+
+Mitigation: set `recursiveReadOnly: IfPossible`.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Core API
+Add `RecursiveReadOnly: (Disabled|IfPossible|Enabled)` to the [`VolumeMount`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/pkg/apis/core/types.go#L1854-L1880) struct:
```

Comment:
I want to try to restate this, and you tell me if you think I captured it properly:

1) If `readOnly` is false, `recursiveReadOnly` must be nil
2) If `readOnly` is true, `recursiveReadOnly` must be set to  one of:
  a) `Disabled`: makes the mountpoint non-recursively read-only
  b) `IfPossible`: makes the mountpoint always read-only, but makes it recursively read-only only if possible.
  c) `Enabled`: makes the mountpoint recursively read-only if possible, otherwise raise an error (as long as the node components are aware of this feature).

Note that point (1) forces a second decision.

We can require that REST logic handle the case of old clients (who do not know about RRO) which change `readOnly` from true to false, while `recursiveReadOnly` is set.  This means that anyone who embeds a PodSpec (e.g. Deployment, ReplicaSet, StatefulSet, Job, ...) has to handle it.  There is no automation for declaring this relationship.  Or...

We can do nothing and let embedded PodSpecs hold data that would be invalid to create a pod.  They will fail to create pods and users will have to figure out why.  This would argue for a common prefix on these field names, so they are always next to each other.

It may be simpler to loosen (1) to:

1') When `readOnly` is false, `recursiveReadOnly` will be ignored, and may be nil or any valid value.

This still leaves a bit of a time-bomb if someone edits and then edits it back.  We probably want an API warning if you try to create a pod with `readOnly: false` and `recursiveReadOnly: <non-nil value>`.

Unfortunately we do not have type-delegated update mechanism like we do for defaults (and that has its own problems).

This whole scenario seems like a corner case of a corner case, so it's unlikely to be terribly consequential, but we do need to define it.  @msau42 @liggitt

Reply:
Just to emphasize:

1) User creates a Deployment with `spec.template.spec.containers[].volumeMounts[].readOnly: true` and `...recursiveReadOnly: Enabled`
2) Old client updates the Deployment and sets `spec.template.spec.containers[].volumeMounts[].readOnly: false`, but doesn't know the other field
3) current state is `readOnly: false`, `recursiveReadOnly: Enabled`

Reply:
We could write a library like `UpdatePodSpec(prev, next *corev1.PodSpec)` and call it from all of the workload controllers Update paths, like we do for validation.

That function could hold the logic like:

```
if *prev.ReadOnly == true && *next.ReadOnly == false {
    // we are converting from RO to RW
    if next.RecursiveReadOnly != nil {
        // RRO is set
        if *next.RecursiveReadOnly == *prev.RecursiveReadOnly {
            // The value was unchanged, so assume the user meant to wipe it.
            next.RecursiveReadOnly = nil
        }
        // else the user set RW *and* changed RRO - let validation catch it
    }
}
```

Reply:
Maybe we do not need to care about that scenario too much, as long as the feature gates stay alpha/beta for enough time?

Reply:
No, we will ALWAYS need to handle this, even in GA.  We have two fields that have a relationship to each other.  They are not strictly independent.  So we need to define what happens in the weird corner cases.

Reply:
The method would only work for controllers we write, but not others that also leverage PodTemplate. But I suppose that is true for any validation of custom resources using PodTemplate.

Reply:
That's right.  If you embed a podTemplate in your CRD today, you are on your own.  If you embed it via an extension API server you can't call our validation code or REST (not exported) either.

Reply:
This is discussed in the [Unions KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1027-api-unions): "...when a client is unaware of new fields added to a union and thus doesn't know how to clear these new fields when trying to set a different field." The KEP appears to conclude that the solution is to introduce a normalization process for unions that "...clears all fields except the one pointed to by the discriminator..."

Unfortunately, because the Unions KEP is not implemented, there is no implementation of the described a normalization process.

Potential things we could do in the near term:

1. Validate that CREATE requests that set RecursiveReadOnly adhere rules that RecursiveReadOnly may only be set when ReadOnly is true.
2. Validate that UPDATE requests that change the value of RecursiveReadOnly also adhere to the rule.
3. Use rest.Updater to clear RecursiveReadOnly on UPDATE when RecursiveReadOnly is present and ReadOnly is changed to false?

Step (3) is the one I'm the least comfortable with.  I'd want to know that this interacts with SSA correctly. WDYT @apelisse ?

If we can't do (3), we could at least send a warning to the client when orphaning a RecursiveReadOnly value and when a client sets ReadOnly to true and RecursiveReadOnly is already set.

cc @deads2k

Reply:
How about this:

1) leave `readOnly` as (non-pointer) bool
2) add RRO as enum-string with default value when `readOnly` is true
3a) on CREATE: if `readOnly` is false and RRO is set, error
3b) on UPDATE: if `readonly` is false and RRO is set, warn

Clients who want to set RRO must set both fields (document this in the API comments).

This lets it get into alpha this cycle and we can discuss whether it is horrible without time pressure.

Then, async:

I'm considering whether we should allow automatic dropping of fields when irrelevant (like @jpbetz suggest - treat this like a union).  Specifically, `readOnly` is the union disciriminator.  If it is false, there one set of config fields (none today) and if it is true there's another set.

Reply:
oh, you know what, I think we've made this more complicated than it needs to be.

First and important point: we have a moratorium on new defaulted fields in PodSpec (https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go#L208-L212).  We can apply defaults when it is actually a pod being created (as opposed to a PodSpec embedded in a Deployment, Job, etc), but his needs to factor into the thinking, which I admit I wasn't doing.

Summary:

Since we can't apply defaults in PodSpecs under Deployment (and Job, and ...) AND these fields under Pods are not mutable, it's actually not so complicated.

When it is a Pod being created, we can default `rro` to a value derived from `ro`, and validate:
* if `ro` is false, `rro` must be nil
* if `ro` is true, `rro` must have a value

When it is a Deployment (or Job, or ...) we cannot default `rro` at all.  This means that any value we find there must have been from the user.  As such, it's fair game for us to validate.  So the only edge-case we need to handle is if:

* `ro` is true; AND
* `rro` is not specified; AND
* the topmost resource is not a Pod

...then we need to allow it.

Does that fly?

Reply:
Why can't we just allow `ro == true && rro == nil` for topmost Pods too?

Reply:
Why is that valuable?  Does it mean something different?  We don't want two ways to express the exact same thing. Better if we can assert that rro will always be set when ro is true. The mechanics of it are easy once you split pods from deployments

Reply:
Interestingly, `MountPropagation` does not seem to have defaulting code, but it should.